slsim/Halos/halos_lens.py

```diff
@@ -0,0 +1,1896 @@
+    is based on the angular diameter distance, which is dependent on the adopted
+    cosmological model. This is particularly useful in cosmological simulations and
+    observations where the physical scale of structures is inferred from angular
+    measurements.
+
+    :param radius_rad: The half cone angle in radians.
+    :param z: The redshift at which the physical area is calculated.
+    :param cosmo: The cosmology instance used for the conversion.
+    :type radius_rad: float
+    :type z: float
+    :type cosmo: astropy.cosmology instance
+    :return: The physical area in square megaparsecs (Mpc^2) for the given cone radius
+        and redshift.
+    :rtype: float :note: The calculation incorporates the angular diameter distance,
+        highlighting the interplay between angular measurements and physical scales in
+        an expanding universe.
+    """
+
+    physical_radius = cosmo.angular_diameter_distance(z) * radius_rad  # Mpc
+    area_physical = np.pi * physical_radius**2
+    return area_physical  # in Mpc2
+
+
+def concentration_from_mass(z, mass, A=75.4, d=-0.422, m=-0.089):
+    """Determine halo concentration from mass using Childs et al. 2018 model.
+
+    This function calculates the concentration parameter of dark matter halos based on
+    their masses, employing the concentration-mass relation presented in Childs et al.
+    2018. This relation is applicable to both relaxed and unrelaxed halos, providing a
+    broad utility for modeling halo properties across different cosmic structures.
+
+    :param z: Redshift of the halo, influencing the concentration-mass relation.
+    :param mass: Mass of the halo in solar masses, serving as the primary input for the
+        concentration calculation.
+    :param A: Pre-factor in the concentration-mass relation, defaulting to 75.4 as per
+        the reference study.
+    :param d: Exponent for the redshift term in the concentration-mass relation, with a
+        default value of -0.422.
+    :param m: Exponent for the mass term in the concentration-mass relation, set to
+        -0.089 by default.
+    :type z: float
+    :type mass:  or array_like
+    :type A: float, optional
+    :type d: float, optional
+    :type m: float, optional
+    :return: The concentration parameter c_200 of the halo(s), providing a measure of
+        the halo's density profile steepness.
+    :rtype: float or array_like :note: The concentration parameter, defined as c_200, is
+        crucial for understanding the density distribution within halos. It is bounded
+        to be at least 1, ensuring physical realism in halo models. :reference: Childs
+        et al. 2018, arXiv:1804.10199, doi:10.3847/1538-4357/aabf95
+    """
+
+    if isinstance(mass, (list, np.ndarray)) and len(mass) == 1:
+        mass = mass[0]
+    c_200 = A * ((1 + z) ** d) * (mass**m)
+    c_200 = np.maximum(c_200, 1)
+    return c_200
+    # TODO: Make this able for list
+
+
+class HalosLens(object):
```

Comment:
I feel this class has too many different roles. Can you split this in different classes with more narrow applications? I.e. 
- one class for the drawing of haloes for a single realization
- one class for the ray-tracing
- one class for the calculation of the mass sheet corrections
- one class to get sample statistics
- perhaps even more classes.
It's hard to review and having so many functionalities in one file can hide a lot of potential duplications for the future